Rebalances pepperbox, cap-and-ball ammunition

DangerNoodle opened the pull request:
Closes https://github.com/CleverRaven/Cataclysm-DDA/issues/18483

Supercedes https://github.com/CleverRaven/Cataclysm-DDA/pull/18549

It has been mentioned that the cap-and-ball revolvers could stand to gain some rebalancing to make them worthwhile compared to other pistol calibers. The first focus is in damage, based off rather time-consuming math and attempts to scavenge for load data (all of rather questionable quality) to allow fudging the numbers in an acceptiable manner.

.36 cap-and-ball damage increased from 17 to 19. This changes it from 1 point more than .32 ACP to directly in-between .32 ACP and .38 Special. Disperson reduced from 120 to 80, making it in between modern equivalents and flintlock ammo.

.44 cap-and-ball damage increased from 26 to 27. This is only enough to just barely make it 1 damage higher than .40 S&W. Disperson reduced from 120 to 70.

Reload times for M1861 Navy and M1860 Army increased by 10 each (to 140 and 160, resperctively). Reload time of pepperbox reduced to in-between these new values. For reference, the flintlock pistol has a reload time of 600.

Additionally ported the weapons and ammo to the correct places, and changed the pepperbox to use .36 cap and ball.

DangerNoodle commented:
Running all of these migrated files through the web linter apparently missed something.

BorkBorkGoesTheCode commented:
Not sure if worth it, but you might add "cotton" to the ammo's item materials. This may allow sewing to repair the ammunition however.

DangerNoodle commented:
That would be strange. Why not paper, since that is what the paper cartridges would logically be make from?

BorkBorkGoesTheCode commented:
Correct.
I noticed that, and started editing my comment before you commented.

BorkBorkGoesTheCode commented:
It looks like the Colt pistols now have one of the best accuracy stats of any ammo-pistol combination in the game. Most others have dispersion ratings in the 300-400 range.

DangerNoodle commented:
That is peculiar. I had intended to set ammunition dispersion in between that of flintlock ammo and comparable pistol ammo.

I assume this means the dispersion of the revolvers themselves warrant a rebalance?

DangerNoodle commented:
It appears as though basic pistol dispersion is way higher than I recall it being.

DangerNoodle commented:
I see. So far I went with 400 for the M1860, and 450 for the M1861.

BorkBorkGoesTheCode commented:
Now that is peculiar, what is wrong with water?

mugling commented:
I did a bad merge earlier, sorry about that. Jenkins is still rebuilding all the PR's

mugling commented on data/json/items/ammo/36paper.json:
```diff
@@ -0,0 +1,21 @@
+    "weight": 6,
+    "volume": 1,
+    "price": 5000,
+    "material": [ "steel", "powder" ],
```
Why `steel`? should that not be `"materials": [ "paper", "powder", "lead" ]`

mugling commented on data/json/items/ammo/44paper.json:
```diff
@@ -0,0 +1,21 @@
+    "weight": 7,
+    "volume": 1,
+    "price": 5000,
+    "material": [ "steel", "powder" ],
```
Same here etc.

mugling commented on data/json/items/gun/44paper.json:
```diff
@@ -0,0 +1,25 @@
+[
+  {
+    "id": "colt_army",
+    "type": "GUN",
```
`copy-from: pistol_revolver` and then drop the redundant fields

DangerNoodle replied:
Seems overriding is easier and, in that case, saves on text. I am assuming the revolver template has a default reload of 70? What about dispersion?

BorkBorkGoesTheCode replied:
>How do I remove mod slots from a copied weapon, for example with the pepperbox?

Look at the Steyr AUG in 223.json: You override them. I think mugling meant that fields like "symbol" "skill" and "color" can be removed, since they are specified by "revolver".

DangerNoodle replied:
As I see. I suppose then the current reload and dispersion can be retained, especially as `proportional` does not get exact reload changes.

mugling replied:
You probably don't need `reload` but you will need `dispersion`

DangerNoodle replied:
What is the default disperion for the revolver template then?

mugling replied:
There isn't one. You can find the abstracts in `data/json/items/classes`

DangerNoodle replied:
Hmm. It would be a bit odd for the reload times of a cap-and-ball pistol to be identical to the template, but it is fine if desired.

mugling replied:
Actually no you're right, perhaps `proportional: 2.0` for twice as long?

DangerNoodle replied:
For all three might work, since 150 and 160 do not make for sane multipliers.

mugling commented on data/json/items/gun/36paper.json:
```diff
@@ -0,0 +1,47 @@
+  },
+  {
+    "id": "pistol_pepperbox",
+    "type": "GUN",
```
`copy-from: pistol_revolver` and then drop the redundant fields

mugling commented on data/json/recipes/ammo/pistol.json:
```diff
@@ -1,5 +1,41 @@
+    "time": 15000,
+    "book_learn": [ [ "manual_pistol", 3 ], [ "recipe_bullets", 3 ] ],
+    "using": [ [ "ammo_bullet", 30 ] ],
+    "qualities": [ { "id": "CUT", "level": 1, "amount": 1 } ],
```
`amount: 1` is always redundant

mugling commented on data/json/recipes/ammo/pistol.json:
```diff
@@ -1,5 +1,41 @@
+    "using": [ [ "ammo_bullet", 30 ] ],
+    "qualities": [ { "id": "CUT", "level": 1, "amount": 1 } ],
+    "components": [
+      [ [ "gunpowder", 100 ], [ "chem_black_powder", 200 ] ],
```
New recipes should be per-charge, especially for this sort of ammunition. Check out the definitions for 9mm for a relevant example

mugling commented on data/json/recipes/ammo/pistol.json:
```diff
@@ -1,5 +1,41 @@
+    "components": [
+      [ [ "gunpowder", 160 ], [ "chem_black_powder", 320 ] ],
+      [ [ "lgpistol_primer", 25 ] ],
+      [ [ "bag_plastic", 1 ], [ "paper", 12 ] ]
```
paper or plastic bag should probably be an `ammo_paper` using group to ensure the ratio's remain consistent between recipes. You could also add aluminum foil in there as well presumably.

DangerNoodle replied:
Any suggested ratio of aluminum foil to paper?

mugling replied:
Probably 12 also (divides a lot of ways)

DangerNoodle replied:
Though this does complicate making them individually, as we can not make it use half a paper or foil.

mugling commented on data/json/recipes/ammo/pistol.json:
```diff
@@ -1,5 +1,41 @@
+    "skill_used": "fabrication",
+    "difficulty": 3,
+    "skills_required": [ "gun", 1 ],
+    "time": 20000,
```
Don't forget to scale time when you change these to per-charge recipes

DangerNoodle commented:
Very well then. I will need to work on these when I can get on later today, as will be leaving soon.

DangerNoodle commented:
Apologies for the delay, can implement changes as proposed by @mugling. Uncertain why commit has retroactively failed.

DangerNoodle commented:
And it seems implementing the suggested "paper cases as a requirement" has unintended consequences, merge conflict.

BorkBorkGoesTheCode commented:
@DangerNoodle : do you know how to revert commits?

DangerNoodle commented:
Yes, but this change should not be causing a conflict requiring reversion. I even updated the requirements file with the latest version before adding the change to it.

I suspect that a requirement entry just for the paper in paper cartridges is more hassle than it is worth.

DangerNoodle commented:
And the linting failures strike again.

mugling commented:
> And the linting failures strike again.

Run your JSON through the web linter

mugling commented:
There's pretty much no reason for `pistol_pepperbox` to exist if you change it's ammo type as that makes it a duplicate of `colt_navy`. It's also pretty badly balanced as `flintlock` so consider dropping it entirely via `data/json/items/migrations.json` via `replace` to `colt_navy`. You then need to drop all the spawns for it. If you're not sure about that I can push that change for you.

Additionally you don't need to specify `flags` when you inherit from `pistol_revolver`. One of the purposes of class-based design is that you don't have to worry about getting the correct set of flags or them later becoming outdated

DangerNoodle commented:
If you are certain that it is acceptable, then some be it. It does resolve the related issue by the strangest of methods (removing the unbalanced item entirely) though.

As for flags, my reasoning was that the revolvers have different combination of flags. They don't eject on reload due to lack of casings (though I would hope this combination would cause no problems), and they are able to jam for some unknown reason. If you are certain that defining reload-ejection does not interact oddly with ammo that lacks casings, then I can make them use the same flags as the revolver template.

mugling commented:
> If you are certain that defining reload-ejection does not interact oddly with ammo that lacks casings, then I can make them use the same flags as the revolver template.

Test it, but I see no reason why it would be a problem

DangerNoodle commented:
Starting it up to test that now, will commit that removal of flags if no problems arise.

mugling commented:
Ok, once your done I can either do the pepperbox migration for you (although if so take not how it happens) or otherwise you can try yourself

DangerNoodle commented:
No apparent issues, so that is not a problem. I have seen the migration files, so the only possible problem that might arise is if changing it causes a merge conflict, in the same way that my attempt to add a requirement for paper casings did.

DangerNoodle commented:
Added migration entry for the pepperbox, removed it from the 36paper gun file, verified that there were no recipes for crafting it, verified that it did not show up in the main itemgroup folder, removed it from the only file in the itemgroup folder that it appears in, verified that it did not show up in any mapgen files.

Did I miss any steps in that regard?

mugling commented:
> Did I miss any steps in that regard?

That sounds like everything. Did you run `--check-mods`?

DangerNoodle commented:
That I forgot. I am unsure how to run that.

mugling commented:
Presuming your on windows you could try via `cmd.exe`

DangerNoodle commented:
Ah right. Need to find where to `cd` to to use it.

DangerNoodle commented:
Searching the mods manually at the moment. Seems I will need to change Generic Guns, and further going over other mods. Even with Notepad++, it takes a while.

DangerNoodle commented:
And here we go with the merge conflicts.

mugling commented:
What exactly are you doing to result in merge conflicts with yourself?

mugling commented:
> And here we go with the merge conflicts.

It's a trivial fix so I'll do it when I merge

DangerNoodle commented:
Ah, thank you. Apologies for that.

mugling commented:
It's not your fault. The `36paper` and `44paper` ammo should probably have `MUZZLE_SMOKE` considering they are blackpowder weapons?

mugling commented:
Good work, those items have gone from neglected/mediocre to having their own niche. You could expand further upon your matchbox primers idea next?

DangerNoodle commented:
I am not sure. I noticed that the amount of smoke given off by the flintlock weapons is always at max intensity, enough to always induce coughing without at least a filter mask. I am unsure if it is a sensible thickness for a muzzleloading rifle, but I do know that it is a bit excessive for a percussion-cap pistol.

DangerNoodle commented:
And it might be a good step to implement match primers, yes. I would also want to double-check another recipe I have seen in other videos, involving mixture of match materials with an additional commerically-available material.

mugling commented:
Play around with `emit.json`, specifically `emit_smoke_plume`

DangerNoodle commented:
Might be interesting to look into it. What would be a suggested density, thin or medium?

mugling commented:
I've just tried density 2 and it stops the coughing (good) but doesn't obscure vision (bad). We don't really want to get into multiple types of muzzle smoke so we need to find balance suitable for most blackpowder weapons.

DangerNoodle commented:
Pecucilar. We do not have a density of smoke capable of affecting vision without provoking excess coughing? Granted, this might entail  partially-obscuring behavior like "reduce vision range in this direction" that is, as far as I am aware, beyond the ability of the game at present.

mugling commented:
I've found the code in the lightmap. I'll look at it more tomorrow so for now blackpowder remains a little on the smoky side.

DangerNoodle commented:
Very well then. This does get me thinking, what is the purpose of the different types of smoke fields for different drugs? Are they meant only to be a visual cheat for a "your character can tell it smells like this type of smoke" mechanic, do they induce effects from secondhand exposure, or are they meant to implement smoke that can obscure vision without provoking coughing?

If it is the last of those, making harmless smoke generic could be one way around this.

mugling commented:
Incidentally your PR set the defacto ratio of blackpowder to gunpowder as 2:1

DangerNoodle commented:
Odd. I was only following earlier examples of ammo using either powder option. I actually was considering changing the ratio to a more practical one, such as 1.5:1.

I am unsure how to calculate an exact recommend difference in powder, as it is hard to find handload data for old cartridges that have both blackpowder and smokeless powder loads. Those that I have found are harder to work with due to the volumetirc method of measuring blackpowder.
